_posts/2023-11-30-prevent-broker-scale-down-if-containing-paritition-replicas.md

```diff
@@ -0,0 +1,181 @@
+
+Then we will deploy the Kafka cluster with Cruise Control enabled.
+
+Example Kafka resource with Cruise Control enabled:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    version: 3.6.0
+    replicas: 4
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.6"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    topicOperator: {}
+    userOperator: {}
+  cruiseControl: {}
+```
+
+Once the cluster is up, we can create some topics such that we have some partition replicas assigned to the brokers
+This is a topic configuration example:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 3
+  replicas: 3
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+```
+
+You can now check how topic partition replicas are allocated across the brokers:
+```shell
+Topic: my-topic	TopicId: bbX7RyTSSXmheaxSPyRIVw	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
+	Topic: my-topic	Partition: 0	Leader: 2	Replicas: 2,3,1	Isr: 2,3,1
+	Topic: my-topic	Partition: 1	Leader: 3	Replicas: 3,1,0	Isr: 3,1,0
+	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
+```
+using the command:
+```shell
+kubectl run -n myproject client -itq --rm --restart="Never" --image="quay.io/strimzi/kafka:latest-kafka-3.6.0" -- \
+sh -c "/opt/kafka/bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --describe --topic my-topic; exit 0";
+```
+
+### Scaling down without moving out all replicas from the broker to be removed
+
+Let's try to scale down `.spec.kafka.replicas` (no. of brokers) from 4 to 3, while all brokers have partition replicas assigned.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    version: 3.6.0
+    replicas: 3            # Changes the replicas to 3
+    listeners:
+# ....    
+```
+
+Now you can apply the updated Kafka custom resource and check if the broker are scaled down or not.
+
+You check the status of the Kafka CR using this command:
+```sh
+kubectl get kafka my-cluster -n myproject -o yaml
+```
+
+Since we didn't move the replicas from the broker to be removed, the scale down will fail and, you will be able to see these errors in the status of the Kafka custom resource
```

Comment:
```suggestion
Since we didn't move the replicas from the broker to be removed, the scale down fails and, as indicated by the errors given in the status of the Kafka custom resource:
```